### PR TITLE
(#509) Fix off-by-1 issues re. MAX_THREADS in clockcache.c

### PR DESCRIPTION
**Pull request by gapisback**
This commit fixes off-by-1 issues checking MAX_THREADS value in assertions in `clockcache.c` . These problems are easily reproduced with `driver_test splinter_test ` with max # of threads. A new test execution with --num-insert-threads 63 is added to nightly runs, in test.sh .

----

The issue #509 fully describes the repro and assertions seen.

These problems were discovered during an in-flight integration while exercising `MAX_THREADS` threads. These fixes were applied to the dev-branch and allowed testing to continue.

Integrating these to /main will help improve overall stability.

### CODE REVIEW

File: src/clockcache.c [hunk 1]
```diff
@@ -863,16 +863,41 @@ clockcache_inc_ref(clockcache *cc, uint32 entry_number, threadid counter_no)
    debug_assert(refcount != MAX_READ_REFCOUNT);
 }
 
+/*
+ * Caller-macro to figure out who's calling this dec_ref interface.
+ *
+ * clockcache_dec_ref(clockcache *cc, uint32 entry_number, threadid counter_no)
+ */
+#define clockcache_dec_ref(cc, entry_number, counter_no)                       \
```
Comment (gapisback) [inline src/clockcache.c:871]: This change isn't directly related to this fix, per se, but if you invoke:

`driver_test splinter_test --perf --num-insert-threads 63 --num-lookup-threads 63 ...`

We will run into an assertion on L893, complaining about ref-count going to 0 after a dec-ref.

I need to log a diff issue to trace this, and so this instrumentation is added to help debug that separate issue.

File: src/clockcache.c [hunk 2]
```diff
@@ -918,7 +943,7 @@ clockcache_assert_no_refs(clockcache *cc)
 {
    threadid        i;
    volatile uint32 j;
-   for (i = 0; i < (MAX_THREADS - 1); i++) {
+   for (i = 0; i < MAX_THREADS; i++) {
```
Comment (gapisback) [inline src/clockcache.c:946]: 1st instance of an off-by-1 check.

File: src/clockcache.c [hunk 3]
```diff
@@ -1354,7 +1379,7 @@ clockcache_batch_start_writeback(clockcache *cc, uint64 batch, bool is_urgent)
 
    clockcache_entry *entry, *next_entry;
 
-   debug_assert(tid < MAX_THREADS - 1);
+   debug_assert((tid < MAX_THREADS), "Invalid tid=%lu\n", tid);
```
Comment (gapisback) [inline src/clockcache.c:1382]: 2nd instance of an off-by-1 check.

File: src/task.c [hunk 4]
```diff
@@ -91,10 +91,11 @@ task_init_threadid(task_system *ts)
    }
 
 out:
+   // The main program counts for thread-0; hence limit is "-1"
    platform_assert((tid != INVALID_TID),
                    "Cannot create a new thread as the limit on"
                    " concurrent threads, %d, will be exceeded.\n",
-                   MAX_THREADS);
+                   (MAX_THREADS - 1));
```
Comment (gapisback) [inline src/task.c:98]: Slightly unrelated fix. 

Without this fix, we will be reporting an error like so, when  `--num-insert-threads 64` is used:

```
Assertion failed at src/task.c:95:task_init_threadid(): "(tid != INVALID_TID)". Cannot create a new thread as the limit on concurrent threads, 64, will be exceeded.
```

The real-limit (externally) is 63 (user + bg-threads)

File: test.sh [hunk 5]
```diff
@@ -270,6 +270,21 @@ function nightly_sync_perf_tests() {
...
+    # Exercise a case with max # of insert-threads which tripped an assertion
+    # This isn't really a 'perf' test but a regression / stability test exec
+    nins_t=63
+    nrange_lookup_t=0
+    test_descr="${nins_t} insert threads"
+    dbname="splinter_test.max_threads.db"
+
+    run_with_timing "Performance with max-threads ${test_descr}" \
+            "$BINDIR"/driver_test splinter_test --perf \
+                                                --num-insert-threads ${nins_t} \
```
Comment (gapisback) [inline test.sh:283]: Adding this case to nightly runs. Even though they run on VMs w/  8 CPUs, we don't particularly care how long this will take to run.

Also, use of `--tree-size-gib 1` reduces # of rows to 1M , which allows the test to complete in a reasonable amount of time.

File: src/clockcache.c [hunk 6]
```diff
@@ -808,16 +808,41 @@ clockcache_inc_ref(clockcache *cc, uint32 entry_number, threadid counter_no)
...
+
    counter_no %= CC_RC_WIDTH;
    uint64 rc_number = clockcache_get_ref_internal(cc, entry_number);
-   debug_assert(rc_number < cc->cfg->page_capacity);
+   debug_assert((rc_number < cc->cfg->page_capacity),
+                "[%d] Entry number, %lu, is out of allocator "
+                "capacity range, %u.\n",
+                line,
+                rc_number,
+                cc->cfg->page_capacity);
```
Comment (ajhconway) [inline src/clockcache.c:834]: I like these more verbose assertions, but please take out the line numbers.

File: src/clockcache.c (hunk 2 above)
Comment (ajhconway) [reply src/clockcache.c:891]: The history of these is that we used to re-use the `(MAX_THREADS - 1)` counter for pins.

File: src/clockcache.c (hunk 1 above)
Comment (ajhconway) [reply src/clockcache.c:816]: I don't see the assertion you're talking about on L893 either in the before or after of this PR.

**Review by ajhconway — CHANGES_REQUESTED** [review]
This is a good catch, but please remove the debugging code.

File: src/clockcache.c (hunk 1 above)
Comment (gapisback) [reply src/clockcache.c:816]: Reverted the debug code as suggested. New diffs to be uploaded soon.

I don't recall which assertion was tripping. I should have documented it better here when I ran into it.
I'll try to repro this later on ...

File: src/clockcache.c [hunk 7]
```diff
@@ -808,16 +808,41 @@ clockcache_inc_ref(clockcache *cc, uint32 entry_number, threadid counter_no)
...
+   clockcache_dec_refcount(cc, entry_number, counter_no, __LINE__)
+
 static inline void
-clockcache_dec_ref(clockcache *cc, uint32 entry_number, threadid counter_no)
+clockcache_dec_refcount(clockcache *cc,
+                        uint32      entry_number,
+                        threadid    counter_no,
+                        int         line)
 {
+   debug_only threadid inp_counter_no = counter_no;
```
[the remark this answers is not in the document]
Comment (gapisback) [reply src/clockcache.c:825]: Done.

**Review by ajhconway — APPROVED** [review]

**Review by rtjohnso — APPROVED** [review]
LGTM
